projecttacoma/measure-repository
# Draft existing artifacts

# Summary
This PR allows the user to pull an existing artifact (Measure or Library) from the measure repository service and clone it with a "draft" status instead of "active" status. This adds this new cloned artifact to the local database.

## New behavior
Before, the user could only create new artifacts from scratch and add it to the local database. Now the user can do so using an existing artifact on the measure repository service. The user can do this in the application in one of two ways: navigate to an existing artifact and click "Clone to Draft Artifact" OR go to the authoring page and under "Start From Existing Measure" the user can select a resource to clone.

## Code changes
- `app/src/pages/[resourceType]/[id].tsx` - new button on a specified measure repository service resource page to clone the artifact to a draft.
- `app/src/pages/api/trpc/[trpc].ts` - increased the NextJS body parser limit (errored out when I tried to clone the MATGlobalCommonFunctionsFHIR4 Library).
- `app/src/pages/authoring/[resourceType]/[id].tsx` - fills the inputs with the correct url and identifier when an artifact is cloned.
- `app/src/pages/authoring/index.tsx` - Added a dropdown on the authoring page for a user to clone an existing artifact.
- `app/src/server/trpc/routers/service.ts` - added two new endpoints: `getArtifactsByResource` (returns all of the artifacts of a certain resourceType) and `convertArtifactById` (takes in a resourceType and an id and calls `createDraft` on that artifact).

# Testing guidance
- `npm run check:all`
- Try cloning Library and Measure artifacts in the MRS through the two ways- directly from the artifact and through the dropdown on the authoring page. Things to look out for:
   - Make sure the number of Library and Measure artifacts increases in the Authoring page nav bar.
   - Go into studio 3T and make sure the artifacts are indeed in the local database once they are cloned.
   - Make sure the status of the cloned artifact is `draft` and not `active`.
   - Make sure that the `url` and `identifier` inputs are populated with the correct values once an artifact is cloned.
   - Make sure the success/error notifications work and are correct based on whether a resource was created or cloned.
